libraries/botbuilder/src/skills/skillHandlerImpl.ts

```diff
@@ -0,0 +1,212 @@
+        } else if (!skillConversationReference.conversationReference) {
+            throw new Error('conversationReference not found.');
+        }
+
+        return skillConversationReference;
+    }
+
+    private async processActivity(
+        claimsIdentity: ClaimsIdentity,
+        conversationId: string,
+        replyToActivityId: string,
+        activity: Activity
+    ): Promise<ResourceResponse> {
+        let resourceResponse: ResourceResponse;
+
+        await this.continueConversation(claimsIdentity, conversationId, async (context, ref) => {
+            /**
+             * This callback does the following:
+             *  - Applies the correct ConversationReference to the Activity for sending to the user-router conversation.
+             *  - For EndOfConversation Activities received from the Skill, removes the ConversationReference from the
+             *    ConversationIdFactory
+             */
+
+            const newActivity = TurnContext.applyConversationReference(activity, ref.conversationReference);
+
+            context.activity.id = replyToActivityId;
+            context.activity.callerId = `${CallerIdConstants.BotToBotPrefix}${JwtTokenValidation.getAppIdFromClaims(
+                claimsIdentity.claims
+            )}`;
+
+            // Legacy BotFrameworkAdapter flows rely on connector client being set inside turn context
+            if (this.adapter instanceof BotFrameworkAdapter) {
+                const client = this.adapter.createConnectorClient(newActivity.serviceUrl);
+                context.turnState.set(this.adapter.ConnectorClientKey, client);
+            }
+
+            switch (newActivity.type) {
+                case ActivityTypes.EndOfConversation:
+                    await this.conversationIdFactory.deleteConversationReference(conversationId);
+
+                    this.applySkillActivityToTurnContext(context, newActivity);
+                    await this.logic(context);
+
+                    break;
+
+                case ActivityTypes.Event:
+                    this.applySkillActivityToTurnContext(context, newActivity);
+                    await this.logic(context);
+
+                    break;
+
+                default:
+                    resourceResponse = await context.sendActivity(newActivity);
+                    break;
+            }
+        });
+
+        if (!resourceResponse) {
+            resourceResponse = { id: uuid() };
+        }
+
+        return resourceResponse;
+    }
+
+    private async continueConversation(
+        claimsIdentity: ClaimsIdentity,
+        conversationId: string,
+        callback: (context: TurnContext, ref: SkillConversationReference) => Promise<void>
+    ): Promise<void> {
+        const ref = await this.getSkillConversationReference(conversationId);
+
+        const continueCallback = (context: TurnContext): Promise<void> => {
+            context.turnState.set(this.skillConversationReferenceKey, ref);
```

Comment:
All flows rely on having skill conversation reference set in turn state, however only `adapter.continueConversation` requires the claims identity to be set in the turn state. Hence the code below.

Reply:
To be more precise, it is set in turn state by the Cloud Adapter (https://github.com/microsoft/botbuilder-js/blob/jpg/cloud-adapter/libraries/botbuilder/src/cloudAdapterBase.ts#L330).